# Topic Modeling with BERTopic

Reproduced from [here](https://maartengr.github.io/BERTopic/index.html)

In [ ]:
# We need to install the library first
# !pip install bertopic

In [1]:
from bertopic import BERTopic
from sklearn.datasets import fetch_20newsgroups

docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']

2025-06-19 16:40:13.378880: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [13]:
# Let's see what the docs look like. This was not in the code but when reproducing the example, it is essential to understand the data.
print(len(docs))
print(type(docs))
print(docs[0])
print(type(docs[0]))
print(len(docs[0]))

18846
<class 'list'>


I am sure some bashers of Pens fans are pretty confused about the lack
of any kind of posts about the recent Pens massacre of the Devils. Actually,
I am  bit puzzled too and a bit relieved. However, I am going to put an end
to non-PIttsburghers' relief with a bit of praise for the Pens. Man, they
are killing those Devils worse than I thought. Jagr just showed you why
he is much better than his regular season stats. He is also a lot
fo fun to watch in the playoffs. Bowman should let JAgr have a lot of
fun in the next couple of games since the Pens are going to beat the pulp out of Jersey anyway. I was very disappointed not to see the Islanders lose the final
regular season game.          PENS RULE!!!


<class 'str'>
712


Great, our data is a *list of strings*. It is a pretty big list with 18846 entries.

In [3]:
# Let's extract topics from the documents

topic_model = BERTopic()
topics, probs = topic_model.fit_transform(docs)

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The 

What did we do here?
- First we initiated an empty `topic_model` of BERTopic class. 
- Then, we called the `fit_transform()` method and pass our `docs` list as the text corpus that the model should be trained on

In [4]:
topic_model.get_topic_info()

,Topic,Count,Name,Representation,Representative_Docs
0,-1,6944,-1_to_the_and_of,"[to, the, and, of, is, for, in, you, it, that]","[In the wake of the Waco denouement, I had ema..."
1,0,1831,0_game_team_games_he,"[game, team, games, he, players, season, hocke...","[\n\n""Deeply rooted rivalry?"" Ahem, Jokerit ha..."
2,1,588,1_key_clipper_chip_encryption,"[key, clipper, chip, encryption, keys, escrow,...",[\nI am not an expert in the cryptography scie...
3,2,530,2_ites_hello_cheek_hi,"[ites, hello, cheek, hi, yep, huh, ken, why, t...","[Hello,, Hello,, ites:]"
4,3,470,3_israel_israeli_jews_arab,"[israel, israeli, jews, arab, jewish, arabs, p...","[\nThis a ""tried and true"" method utilized by ..."
...,...,...,...,...,...
221,220,11,220_eu_powerbook_warranty_amex,"[eu, powerbook, warranty, amex, purchase, citi...","[: Most, if not all, credit card companies off..."
222,221,11,221_hotel_voucher_nights_vacation,"[hotel, voucher, nights, vacation, reservation...","[The package is called Sun and Sand, it includ..."
223,222,11,222_eisa_bus_isa_motherboard,"[eisa, bus, isa, motherboard, board, 32bit, dm...",[\nI don't see this as a problem with the chip...
224,223,10,223_48bit_color_screen_bits,"[48bit, color, screen, bits, sgis, 12bit, info...",[\n[ description of experiment deleted ]\n\n\n...


In [5]:
topic_model.get_topic(0)

[('game', 0.010223588055360033),
 ('team', 0.008914638758091042),
 ('games', 0.007105900091394138),
 ('he', 0.0068619970068052785),
 ('players', 0.0062643815301323446),
 ('season', 0.006184029456545325),
 ('hockey', 0.0060749090677381205),
 ('play', 0.005716857731607741),
 ('25', 0.00557476496472491),
 ('year', 0.005546974883540843)]

In [6]:
topic_model.get_topic(-1)

[('to', 0.004298714769042394),
 ('the', 0.004227723552484535),
 ('and', 0.0041331145598501205),
 ('of', 0.004124600200578983),
 ('is', 0.004087236102973601),
 ('for', 0.003915014761653062),
 ('in', 0.003911188986350228),
 ('you', 0.0038896873053654466),
 ('it', 0.0038616872979240365),
 ('that', 0.0037318774446026405)]

topic -1 is the outlier topic and it should be ignored.

In [7]:
topic_model.get_document_info(docs)

,Document,Topic,Name,Representation,Representative_Docs,Top_n_words,Probability,Representative_document
0,\n\nI am sure some bashers of Pens fans are pr...,0,0_game_team_games_he,"[game, team, games, he, players, season, hocke...","[\n\n""Deeply rooted rivalry?"" Ahem, Jokerit ha...",game - team - games - he - players - season - ...,1.000000,False
1,My brother is in the market for a high-perform...,-1,-1_to_the_and_of,"[to, the, and, of, is, for, in, you, it, that]","[In the wake of the Waco denouement, I had ema...",to - the - and - of - is - for - in - you - it...,0.000000,False
2,\n\n\n\n\tFinally you said what you dream abou...,16,16_armenian_turkish_armenians_genocide,"[armenian, turkish, armenians, genocide, turke...",[\n\nPoor 'Poly'. I see you're preparing the g...,armenian - turkish - armenians - genocide - tu...,0.246081,False
3,\nThink!\n\nIt's the SCSI card doing the DMA t...,53,53_scsi_scsi2_scsi1_ide,"[scsi, scsi2, scsi1, ide, asynchronous, drive,...",[wlsmith@valve.heart.rri.uwo.ca (Wayne Smith) ...,scsi - scsi2 - scsi1 - ide - asynchronous - dr...,0.656768,False
4,1) I have an old Jasmine drive which I cann...,78,78_tape_backup_tapes_drive,"[tape, backup, tapes, drive, munroe, wangdat, ...",[hello all- i have a problem with my micro sol...,tape - backup - tapes - drive - munroe - wangd...,0.876293,False
...,...,...,...,...,...,...,...,...
18841,DN> From: nyeda@cnsvax.uwec.edu (David Nye)\nD...,-1,-1_to_the_and_of,"[to, the, and, of, is, for, in, you, it, that]","[In the wake of the Waco denouement, I had ema...",to - the - and - of - is - for - in - you - it...,0.000000,False
18842,\nNot in isolated ground recepticles (usually ...,164,164_ground_grounding_conductor_neutral,"[ground, grounding, conductor, neutral, wire, ...","[\nNot according to the NEC nor the CEC, as ex...",ground - grounding - conductor - neutral - wir...,0.886714,False
18843,I just installed a DX2-66 CPU in a clone mothe...,89,89_fan_cpu_heat_sink,"[fan, cpu, heat, sink, fans, cooling, chip, co...","[N(P>Just got a 66MHz 486DX2 system, and am co...",fan - cpu - heat - sink - fans - cooling - chi...,0.901844,False
18844,\nWouldn't this require a hyper-sphere. In 3-...,149,149_den_sphere_radius_plane,"[den, sphere, radius, plane, points, circle, s...","[\nHere is one by Andrew ""Graphics Gems"" Glass...",den - sphere - radius - plane - points - circl...,0.976180,False


In [11]:
# Let's save this as a dataframe
import pandas as pd
df = topic_model.get_document_info(docs)
print(df.head())

                                            Document  Topic  \
0  \n\nI am sure some bashers of Pens fans are pr...      0   
1  My brother is in the market for a high-perform...     -1   
2  \n\n\n\n\tFinally you said what you dream abou...     16   
3  \nThink!\n\nIt's the SCSI card doing the DMA t...     53   
4  1)    I have an old Jasmine drive which I cann...     78   

                                     Name  \
0                    0_game_team_games_he   
1                        -1_to_the_and_of   
2  16_armenian_turkish_armenians_genocide   
3                 53_scsi_scsi2_scsi1_ide   
4              78_tape_backup_tapes_drive   

                                      Representation  \
0  [game, team, games, he, players, season, hocke...   
1     [to, the, and, of, is, for, in, you, it, that]   
2  [armenian, turkish, armenians, genocide, turke...   
3  [scsi, scsi2, scsi1, ide, asynchronous, drive,...   
4  [tape, backup, tapes, drive, munroe, wangdat, ...   

             

In [12]:
# Let's look at one example document and its topic distribution
print(docs[0])
print(df["Representation"][0])




I am sure some bashers of Pens fans are pretty confused about the lack
of any kind of posts about the recent Pens massacre of the Devils. Actually,
I am  bit puzzled too and a bit relieved. However, I am going to put an end
to non-PIttsburghers' relief with a bit of praise for the Pens. Man, they
are killing those Devils worse than I thought. Jagr just showed you why
he is much better than his regular season stats. He is also a lot
fo fun to watch in the playoffs. Bowman should let JAgr have a lot of
fun in the next couple of games since the Pens are going to beat the pulp out of Jersey anyway. I was very disappointed not to see the Islanders lose the final
regular season game.          PENS RULE!!!


['game', 'team', 'games', 'he', 'players', 'season', 'hockey', 'play', '25', 'year']


In [9]:
topic_model.visualize_topics()